File: lintcode/589.connecting-graph.cpp
class ConnectingGraph {
private:
    vector<int> table;
public:
    /*
    * @param n: An integer
    */ConnectingGraph(int n) {
        // do intialization if necessary
        for (auto i = 0; i <= n; ++i)
        {
            table.push_back(i);
        }
    }


    int find(int a){
        if (a == table[a])
        {
            return a;
        }
        
        table[a] = find(table[a]);
        return table[a];
    }

    /*
     * @param a: An integer
     * @param b: An integer
     * @return: nothing
     */
    void connect(int a, int b) {
        // write your code here
        int root_a = find(a);
        int root_b = find(b);
        
        if (root_a != root_b)
        {
            table[root_a] = root_b;
        }
    }

    /*
     * @param a: An integer
     * @param b: An integer
     * @return: A boolean
     */
    bool query(int a, int b) {
        // write your code here
        int root_a = find(a);
        int root_b = find(b);
        return root_a == root_b;
    }
};

File: lintcode/589.connecting-graph.py
class ConnectingGraph:
    """
    @param: n: An integer
    """
    def __init__(self, n):
        # do intialization if necessary
        self.dict = {}
        for i in range(1, n + 1):
            self.dict[i] = i

    def find(self, node):
        if self.dict[node] == node:
            return node
        self.dict[node] = self.find(self.dict[node])
        return self.dict[node]

    """
    @param: a: An integer
    @param: b: An integer
    @return: nothing
    """
    def connect(self, a, b):
        # write your code here
        a_root = self.find(a)
        b_root = self.find(b)

        self.dict[a_root] = b_root

    """
    @param: a: An integer
    @param: b: An integer
    @return: A boolean
    """
    def query(self, a, b):
        # write your code here
        return self.find(a) == self.find(b)